[SPARK-38159][SQL] Add a new FileSourceMetadataAttribute for the Hidden File Metadata

Yaohua628 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
Add a new `FileSourceMetadataAttribute` object with an `apply` method to create a `FileSourceMetadataAttribute`, and an `unapply` method to match only file source metadata attribute.


### Why are the changes needed?
Extra safeguard to make sure it matches file source metadata attribute.


### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Existing UTs

Yaohua628 committed:
unapply also matches on attr name

Yaohua628 commented:
Hi @cloud-fan, an XS PR to refactor the `MetadataAttribute` `unapply` method, please take a look whenever you have time, thanks a lot!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -437,16 +437,20 @@ object VirtualColumn {
- * - unapply() will check if an attribute reference is the metadata attribute reference
+ * - unapply() will check if an attribute reference is the file source metadata attribute reference
  */
 object MetadataAttribute {
```
Maybe we should have two objects
1. `MetadataAttribute`: This is used to create and extract general metadata attributes (not only for file source)
2. `FileSourceMetadataAttribute`: To extract file source metadata attributes

cloud-fan replied:
Or we make sure that we only call `MetadataAttribute.unapply` in places that only deal with file sources.

Yaohua628 replied:
make sense!

Yaohua628 committed:
comments

Yaohua628 committed:
nits

Yaohua628 commented:
@cloud-fan addressed your comment by adding a `FileSourceMetadataAttribute`, thanks!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -451,3 +451,21 @@ object MetadataAttribute {
+  val ATTRIBUTE_NAME = "_metadata"
+
+  def unapply(attr: AttributeReference): Option[AttributeReference] = {
+    if (attr.metadata.contains(METADATA_COL_ATTR_KEY)
```
nit:
```
attr match {
  case MetadataAttribute(a) if a.name.equalsIgnoreCase(ATTRIBUTE_NAME) => Some(attr)
  case _ => None
}
```

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:
```diff
@@ -214,12 +214,12 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
       }
 
       val metadataColumns = metadataStructOpt.map { metadataStruct =>
         metadataStruct.dataType.asInstanceOf[StructType].fields.map { field =>
```
now we can do `metadataStruct.dataType.asInstanceOf[StructType].toAttributes`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:
```diff
@@ -366,9 +364,8 @@ case class FileSourceScanExec(
   private lazy val pushedDownFilters = {
     val supportNestedPredicatePushdown = DataSourceUtils.supportNestedPredicatePushdown(relation)
-    // TODO: should be able to push filters containing metadata columns down to skip files
     dataFilters.filterNot(_.references.exists {
```
I think it's better to remove the metadata col filter in `FileSourceStrategy`, as that's where we make the metadata col filter invalid: we remove the struct-type metadata col and convert its fields to attributes.

Yaohua628 replied:
agreed, it is better to remove in `FileSourceStrategy`, but we actually need to use the intact `dataFilters` here in `FileSourceScanExec` to filter files to read (i.e. `listFiles(..., dataFilters)`). 

added some comments here, do you think it is OK for now?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -451,3 +451,21 @@ object MetadataAttribute {
+ */
+object FileSourceMetadataAttribute {
+
+  val ATTRIBUTE_NAME = "_metadata"
```
After more thought, I think it's more robust to not rely on the attribute name to extract file source metadata attrs. We can add a new metadata key to indicate file source metadata attributes. Sorry for the back and forth!

Yaohua628 replied:
np! sounds good to me!

AmplabJenkins commented:
Can one of the admins verify this patch?

Yaohua628 committed:
add new key

Yaohua628 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -451,3 +451,28 @@ object MetadataAttribute {
+ * - apply() will create a file source metadata attribute reference
+ * - unapply() will check if an attribute reference is the file source metadata attribute reference
+ */
+object FileSourceMetadataAttribute {
```
still keep the `METADATA_COL_ATTR_KEY` in `FileSourceMetadataAttribute`: `FileSourceMetadataAttribute` <= `MetadataAttribute`

Yaohua628 commented:
@cloud-fan addressed all your comments: 
- added a new key for `FileSourceMetadataAttribute`
- addressed some nits
- replied one of your suggestions

ready for another pass, thanks!

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:
```diff
@@ -196,12 +196,10 @@ case class FileSourceScanExec(
     tableIdentifier: Option[TableIdentifier],
-    disableBucketedScan: Boolean = false)
+    disableBucketedScan: Boolean = false,
+    metadataColumns: Seq[AttributeReference])
```
Now we can avoid adding this new parameter. When convert the fields of struct-type metadata col, we can use `FileSourceMetadataAttribute`

Yaohua628 replied:
oops, forgot to revert it back...

Yaohua628 committed:
revert: not add new parameter

cloud-fan commented:
document building fails. Can you rerun the Github Action job?

Yaohua628 commented:
> document building fails. Can you rerun the Github Action job?

done! thanks

cloud-fan commented:
thanks, merging to master!
